fix maximum call stack size reached issue

alanrenmsft opened the pull request:
This PR fixes #14744 

using spread operator as method parameter is not safe, it might exceed the maximum stack size when dealing with a lot of records (130k in this case).

I have a PR for vscode and I have cherrypicked the changes in this PR.

kburtram commented on src/sql/workbench/contrib/query/browser/messagePanel.ts:
```diff
@@ -201,7 +201,7 @@ export class MessagePanel extends Disposable {
 	private onMessage(message: IQueryMessage | IQueryMessage[], setInput: boolean = false) {
 		if (isArray(message)) {
-			this.model.messages.push(...message);
+			this.model.messages = this.model.messages.concat(message);
```
does this allocate and throw away an array for each message?  if so, is there any perf impact from the memory churn?

alanrenmsft replied:
not using concat in the new implementation anymore.

kburtram reviewed (approved):
LGTM.  though I'm not sure if this change is managing memory particularly efficiently if I'm reading the code correctly.

alanrenmsft commented:
I thought about it again and the problem is that we need an alternative way to do in place array push and splice that is safe for large arrays, so I implemented those and added it to the vscode arrays.ts. 

@kburtram Array.concat will return a new array which might not work for some scenarios, it might use more memory than the in place push but not a huge concern as we are not holding them and will be GC'ed by the runtime. 

while implementing the push/insert methods, I picked the approach of setting array length and then updating the values since it is faster comparing to other in place alternatives, it becomes more obvious as the number of items grow. here are the perf numbers, the test was done on https://jsbench.me/ 

2 array with 1000 items
![image](https://user-images.githubusercontent.com/13777222/112708598-171a1a00-8e70-11eb-9731-aef42c315d97.png)
2 array with 10,000 items
![image](https://user-images.githubusercontent.com/13777222/112708609-29945380-8e70-11eb-9a2f-51c913959dae.png)
2 array with 100,000 items
![image](https://user-images.githubusercontent.com/13777222/112708621-36b14280-8e70-11eb-97b3-726fd5cb6ce3.png)
2 array with 1,000,000 items
![image](https://user-images.githubusercontent.com/13777222/112708631-42046e00-8e70-11eb-9fae-189a62cab98e.png)

Charles-Gagnon commented on src/vs/base/common/arrays.ts:
```diff
@@ -612,3 +612,78 @@ export function mapFind<T, R>(array: Iterable<T>, mapFn: (value: T) => R | undef
 }
+
+/**
+ * An alternative for Array.push(...items) method, Use this if you need to push large number of items to the array.
```
Would suggest just putting these in our own arrays.ts class unless you're planning on trying to get VS Code to take it

alanrenmsft replied:
I need to use the splice method in a vscode file, here is the vscode PR https://github.com/microsoft/vscode/pull/119929
